Show error messages in create/edit GTAP modal

camsaul opened the pull request:
Fixes (part of?) #14612

GTAP API endpoints weren't actually returning an error message -- just info about the error. GTAP modal on frontend was not displaying errors either way. This PR fixes error messages returned by API endpoints and displays them in the GTAP modal

![image](https://user-images.githubusercontent.com/1455846/106974644-e805e880-6709-11eb-8d77-68be4d7c21ff.png)

Not sure if this should close #14612 entirely... thoughts @flamber ?

camsaul committed:
Rework tests in api.gtap-test

camsaul committed:
Return actual error message when trying to save an invalid sandbox query

camsaul committed:
Minor refactor

camsaul committed:
GTAP modal should display errors

camsaul committed:
Fix flow failure

camsaul committed:
Test fix

camsaul committed:
Another flow fix :wrench:

camsaul committed:
Longer timeout for the test that keeps failing

paulrosenzweig commented on enterprise/frontend/src/metabase-enterprise/sandboxes/components/GTAPModal.jsx:
```diff
@@ -119,10 +123,21 @@ export default class GTAPModal extends React.Component {
+      }
+    } catch (error) {
+      console.error("Error saving GTAP", error);
+      const message = error
+        ? error.data
+          ? error.data.message || JSON.stringify(error.data)
+          : JSON.stringify(error)
+        : UNKNOWN_ERROR_MESSAGE;
```
`FormMessage` should probably export a function that contains this logic.

paulrosenzweig commented on enterprise/frontend/src/metabase-enterprise/sandboxes/components/GTAPModal.jsx:
```diff
@@ -119,10 +123,21 @@ export default class GTAPModal extends React.Component {
+          : JSON.stringify(error)
+        : UNKNOWN_ERROR_MESSAGE;
+      this.setState({ error: message });
+      throw new Error(message);
```
Why rethrow? To avoid closing the modal?

camsaul replied:
yes

paulrosenzweig reviewed (approved):
FE looks good to me!

flamber commented:
Not seeing any messages on CI build of `06eb4a6`. I'm not sure if CI build is not working or I'm not doing the right thing to see the errors.

dpsutton commented on enterprise/backend/test/metabase_enterprise/sandbox/models/group_table_access_policy_test.clj:
```diff
@@ -74,7 +74,7 @@
         (testing "adding new columns = not ok"
```
I'm guessing the fix to this is the inclusion of `{:card_id msg}` in the error. If that's what the UI is expecting in order to hook the error up to the card displayed to the user should we have a test checking that that is present?

camsaul replied:
Fixed. This was supposed to be `:message`. @flamber , can you try again

dpsutton commented on enterprise/backend/src/metabase_enterprise/sandbox/models/group_table_access_policy.clj:
```diff
@@ -67,19 +67,23 @@
-                          :new-column  col
-                          :expected    (mapv :name table-cols)
-                          :actual      (mapv :name result-metadata-columns)})))
+         (let [msg (tru "Sandbox Cards can''t return columns that aren''t present in the Table they are sandboxing.")]
```
`aren''t` double apostrophe

camsaul replied:
You have to use double apostrophes in `tru`/trs`. Single ones are for things that are supposed to be returned literally without being quoted or something like that. I put some error checking in the `trs`/`tru` macros that is supposed to catch this but it missed that one for some reason. I noticed it and fixed it here

dpsutton replied:
ah. cool. my bad. :)

dpsutton commented:
Just the one comment. I'm not sure what the actual fix was here. The error message already had the string "Sandbox Cards can''t return columns that aren''t present in the Table they are sandboxing." in it. But then you added it to the `errors {:card_id msg}` spot.  If the data is the fix we probably should have a test ensuring that data is correctly put there. I'm also a bit confused by the shape. Associated to a key like `:card_id` I would expect an integer not a string.

camsaul committed:
Return error message as :message

camsaul commented:
@dpsutton the convention we've been using to return error messages from the API is *usually* like 

```clj
{:errors {field-with-error message}}
```

But not everything follows that pattern. In this case I thought `card_id` was the Field most clearly associated with the error. I tried adding that first to see if it would make the error message show up. That ended up not working, so I had to go add FE code to show the error as well.

In a few other places the API just returns a string error message e.g. `"Not found"` or returns a `:message` key. Since I was adding the FE code I ultimately went with using `message` instead of `errors`. I fixed the PR to explicitly include that change. I originally made some tweaks to the API middleware to always include the exception message as `:message`, so it *was* working for me when I took that screenshot, but I reverted that middleware change thinking it wasn't needed. So that's probably why it didn't work for @flamber just now.

It should be fixed now however

camsaul commented:
@nemanjaglumac, what are the chances you could do me a humongous favor and write a test case to make sure Sandbox error messages show up now in this PR? I think you'd be able to do it a lot quicker than I would

nemanjaglumac commented:
> @nemanjaglumac, what are the chances you could do me a humongous favor and write a test case to make sure Sandbox error messages show up now in this PR? I think you'd be able to do it a lot quicker than I would

Sure thing.
I'll try to reuse parts of old https://github.com/metabase/metabase-enterprise/issues/524.

dpsutton commented:
@camsaul that makes sense. thanks for the explanation

camsaul committed:
New cache key for drivers/uberjar so Oracle/Vertica errors stop popping up

camsaul committed:
Revert cache key rotation

camsaul commented:
The failing Oracle/Vertica Cypress test should be fixed by #14655

camsaul committed:
Merge branch 'release-x.38.x' into meaningful-error-messages

nemanjaglumac committed:
#14612 Repro: Sandboxing limitations should have meaningful UI error message/feedback (#14690)

camsaul committed:
Fix random Cypress test failures when fingerprinting fails?

camsaul committed:
Bump no-output-timeouts to 15m because CircleCI is extra slow lately

camsaul committed:
Merge branch 'meaningful-error-messages' of github.com:metabase/metabase into meaningful-error-messages

flamber reviewed (commented):
Awesome - gives good error messages to the admin :+1:
Tested commit `a292494`

But if I upgrade from 1.37.8 with a sandbox-card with an extra column, then the users might see this not very helpful error:
`Output of gtap->source does not match schema: [0;33m {:source-metadata [nil nil nil nil nil nil nil nil (not (map? nil))]} [0m `
![image](https://user-images.githubusercontent.com/1447303/107118872-0132a480-6884-11eb-9946-6b3d8d8e00be.png)

---

The admin can still create/change questions in a way where they don't see any nice error messages, so they can still add columns unintentionally.
Basically the same problem as the reopened part of EE520.

Can change an already connected SQL (adding `, 'x' AS "ExtraCol"`), since the query doesn't run because the variable is non-optional (not specifically with Required set):
```
SELECT *, 'x' AS "ExtraCol" FROM "PRODUCTS" WHERE "CATEGORY"={{sandbox}}
```
The user then has the "ExtraCol" in their view too:
![image](https://user-images.githubusercontent.com/1447303/107119084-8f5b5a80-6885-11eb-9504-5f546e008984.png)

It's also possible to create a Notebook question and do stuff (add Custom Column etc) and then save the question without clicking Visualize, but that's an edge case (think it's related to issue 13468).
